### PR TITLE
Add some more semantics to EnzymeConstrainedModel

### PR DESCRIPTION
**Pull request by stelmo**
Semantics are awesome ❤️

### CODE REVIEW

File: src/solver.jl [hunk 1]
```diff
@@ -130,9 +148,7 @@ Returns a vector of fluxes of the model, if solved.
 flux_vector(flux_balance_analysis(model, ...))
 ```
 """
-flux_vector(model::AbstractMetabolicModel, opt_model)::Maybe{Vector{Float64}} =
-    is_solved(opt_model) ? reaction_variables_matrix(model)' * value.(opt_model[:x]) :
-    nothing
+flux_vector(model::AbstractMetabolicModel, opt_model)::Maybe{Vector{Float64}} = get_solution(:reaction, model, opt_model)
```
Comment (exaexa) [inline src/solver.jl:151]: maybe we could just rename or completely eradicate this now? `flux_` sounds unsystematic to `reaction_`. Flux can be anything

File: src/solver.jl [hunk 2]
```diff
@@ -120,6 +120,24 @@ solved_objective_value(flux_balance_analysis(model, ...))
...
+
+function get_solution_dict(semantics::Val{Semantics}, model::AbstractMetabolicModel, opt_model) where {Semantics}
+    sem = Accessors.Internal.get_semantics(semantics)
+    isnothing(sem) && throw(DomainError(semantics, "Unknown semantics"))
+    (ids, _, _, sem_varmtx) = sem
+    is_solved(opt_model) ? Dict(ids .=> sem_varmtx(model)' * value.(opt_model[:x])) : nothing
+end
+
+get_solution_dict(semantics::Symbol, model::AbstractMetabolicModel, opt_model) = get_solution_dict(Val(semantics), model, opt_model)
+
```
Comment (exaexa) [inline src/solver.jl:140]: This is great, thanks.

As a tuning opportunity, the `get_` prefixes implies an imperative extraction operation, which should be possibly avoided, we want the whole thing to feel as stateless as possible. `solution` is probably too short, maybe we go for something like `solution_variable_values` and `solution_reaction_values` etc? Other ideas welcome tho.

File: src/types/wrappers/EnzymeConstrainedModel.jl [hunk 3]
```diff
@@ -181,13 +181,23 @@ the original fluxes in the wrapped model
...
         variables(model),
-        semantics(model),
+        reactions(model),
         reaction_variables_matrix(model),
     ) # TODO currently inefficient
 
 """
 $(TYPEDSIGNATURES)
 
+Get a mapping of enzyme variables to variables - for enzyme constrained models,
```
Comment (exaexa) [inline src/types/wrappers/EnzymeConstrainedModel.jl:191]:
```suggestion
Get a mapping of enzyme variables to variables -- for enzyme constrained models,
```
hyphen vs dash

File: src/types/wrappers/EnzymeConstrainedModel.jl [hunk 4]
```diff
@@ -181,13 +181,23 @@ the original fluxes in the wrapped model
...
+this is just a direct mapping.
+"""
+Accessors.enzyme_variables(model::EnzymeConstrainedModel) = Dict(
+    gid => Dict(gid => 1.0) for gid in genes(model)
```
Comment (exaexa) [inline src/types/wrappers/EnzymeConstrainedModel.jl:195]: oh lovely

File: src/solver.jl [hunk 5]
```diff
@@ -120,6 +120,24 @@ solved_objective_value(flux_balance_analysis(model, ...))
...
     is_solved(opt_model) ? objective_value(opt_model) : nothing
 
+function get_solution(semantics::Val{Semantics}, model::AbstractMetabolicModel, opt_model) where {Semantics}
+    sem = Accessors.Internal.get_semantics(semantics)
+    isnothing(sem) && throw(DomainError(semantics, "Unknown semantics"))
+    (_, _, _, sem_varmtx) = sem
+    is_solved(opt_model) ? sem_varmtx(model)' * value.(opt_model[:x]) : nothing
+end
+
+get_solution(semantics::Symbol, model::AbstractMetabolicModel, opt_model) = get_solution(Val(semantics), model, opt_model)
```
Comment (exaexa) [inline src/solver.jl:130]: for this to be pipeable you need to remove the last argument and return a new closure
```suggestion
get_solution(semantics::Symbol, model::AbstractMetabolicModel) = opt_model -> get_solution(Val(semantics), model, opt_model)
```

File: src/solver.jl [hunk 6]
```diff
@@ -120,6 +120,24 @@ solved_objective_value(flux_balance_analysis(model, ...))
...
+
+function get_solution_dict(semantics::Val{Semantics}, model::AbstractMetabolicModel, opt_model) where {Semantics}
+    sem = Accessors.Internal.get_semantics(semantics)
+    isnothing(sem) && throw(DomainError(semantics, "Unknown semantics"))
+    (ids, _, _, sem_varmtx) = sem
+    is_solved(opt_model) ? Dict(ids .=> sem_varmtx(model)' * value.(opt_model[:x])) : nothing
```
Comment (exaexa) [inline src/solver.jl:136]:
```suggestion
    is_solved(opt_model) ? Dict(ids(model) .=> sem_varmtx(model)' * value.(opt_model[:x])) : nothing
```
`ids` is a getter function

Comment (exaexa) [reply src/solver.jl:136]: (this should probably fix the tests)

**Review by exaexa — COMMENTED** [review]
I'm perfectly OK with this once the tests work (see the comments for hints)

File: src/solver.jl (hunk 2 above)
Comment (stelmo) [reply src/solver.jl:140]: will think of something (currently favoring `value`)

File: src/solver.jl (hunk 1 above)
Comment (stelmo) [reply src/solver.jl:151]: will eradicate this in the next PR

**Review by exaexa — APPROVED** [review]
VERY GOOD.

I stil think we will need to find a good name for the `values_*`.... perhaps even `solution_values_...` ? but that can be left for later.

**Pull request merged**